src/response/error.rs

```diff
@@ -1,70 +1,142 @@
     /// his exception typically implies that the request wasn't sent to the service successfully or if the service responded, the sdk couldn't interpret the response.
     /// An example would be SDK client was unable to convert the user provided data into a valid request that was expected by the service.
-    ClientSdkError(String),
+    #[error("client error: {description}")]
+    ClientSdkError {
+        description: Cow<'static, str>,
+        #[source]
+        source: ErrorSource,
+    },
+
     /// SDK client side validation fails.
-    InvalidArgument(String),
-}
+    #[error("invalid argument: {description}")]
+    InvalidArgument {
+        description: Cow<'static, str>,
+        #[source]
+        source: Option<ErrorSource>,
+    },
 
-impl std::fmt::Display for MomentoError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            MomentoError::InternalServerError(e)
-            | MomentoError::BadRequest(e)
-            | MomentoError::PermissionDenied(e)
-            | MomentoError::Unauthenticated(e)
-            | MomentoError::NotFound(e)
-            | MomentoError::AlreadyExists(e)
-            | MomentoError::Cancelled(e)
-            | MomentoError::Timeout(e)
-            | MomentoError::LimitExceeded(e)
-            | MomentoError::ClientSdkError(e)
-            | MomentoError::InvalidArgument(e) => write!(f, "{e}"),
-        }
-    }
+    /// Requested operation was interrupted.
+    /// This may happen to a Topic subscription, or it may show up due to an HTTP2 GOAWAY graceful reconnection request from the server.
+    /// Whatever the case, you can probably retry the request.
+    #[error("operation interrupted: {description}")]
+    Interrupted {
+        description: Cow<'static, str>,
+        #[source]
+        source: ErrorSource,
+    },
 }
 
-impl std::error::Error for MomentoError {}
+#[derive(Debug, thiserror::Error)]
+pub enum ErrorSource {
+    /// A source you will need to downcast if you need to do something with it.
+    #[error("unknown source")]
+    Unknown(#[from] Box<dyn std::error::Error + Send + Sync>),
 
-impl From<http::uri::InvalidUri> for MomentoError {
-    fn from(e: http::uri::InvalidUri) -> Self {
-        // the uri gets derived from the jwt
-        Self::ClientSdkError(e.to_string())
-    }
-}
+    /// A detailed error from the Auth module
+    #[error("auth error")]
+    AuthError(#[from] AuthError),
 
-impl From<String> for MomentoError {
-    fn from(s: String) -> Self {
-        Self::BadRequest(s)
-    }
+    /// Caused by something in our backing library Tonic
+    #[error("tonic transport error")]
+    TonicTransport(#[from] tonic::transport::Error),
+
+    /// Caused by something in our backing library Tonic
+    #[error("tonic status error")]
+    TonicStatus(#[from] tonic::Status),
+
+    /// Caused by a malformed URI
+    #[error("uri is invalid")]
+    InvalidUri(#[from] http::uri::InvalidUri),
 }
```

Comment:
These should be using `#[source]` instead of `#[from]` right? We probably want to continue the source chain through the `ErrorSource`.

(this was meant to be part of the review above but I forgot to actually create the comment)

Reply:
```
The Error trait’s source() method is implemented to return whichever field has a #[source] attribute or is named source, if any. This is for identifying the underlying lower level error that caused your error.

The #[from] attribute always implies that the same field is #[source], so you don’t ever need to specify both attributes.
```
`from` means `source + from` [according to the documentation](https://docs.rs/thiserror/latest/thiserror/)